storage-bigtable/src/lib.rs

```diff
@@ -790,21 +790,41 @@ impl LedgerStorage {
             })
             .collect();
 
+        let mut tasks = vec![];
+
         if !tx_cells.is_empty() {
-            bytes_written += self
-                .connection
-                .put_bincode_cells_with_retry::<TransactionInfo>("tx", &tx_cells)
-                .await?;
+            let conn = self.connection.clone();
+            tasks.push(tokio::spawn(async move {
+                conn.put_bincode_cells_with_retry::<TransactionInfo>("tx", &tx_cells)
+                    .await
+            }));
         }
 
         if !tx_by_addr_cells.is_empty() {
-            bytes_written += self
-                .connection
-                .put_protobuf_cells_with_retry::<tx_by_addr::TransactionByAddr>(
+            let conn = self.connection.clone();
+            tasks.push(tokio::spawn(async move {
+                conn.put_protobuf_cells_with_retry::<tx_by_addr::TransactionByAddr>(
                     "tx-by-addr",
                     &tx_by_addr_cells,
                 )
-                .await?;
+                .await
+            }));
+        }
+
+        let results = futures::future::join_all(tasks).await;
+        let results: Vec<_> = results.into_iter().map(|r| r.unwrap()).collect();
```

Comment:
```suggestion
        let results: results.into_iter().map(|r| r).collect::<Result<Vec<_>>>()?;
```

Then I think you can clean up the stuff below, right?

Reply:
results is a Vec<Result<Result<usize>> where the first Result is a tokio::runtime::Result and the second one is a bigtable Result. 

i figure the join is _probably_ safe to unwrap, the bigtable result much more likely to fail

Reply:
I believe Trent is suggesting you use the `collect()` to find the first bigtable error, meaning you don't need the block on L825 below.
The bytes_written logic would have to come into the iterator, of course.

I'm torn on whether to unwrap the JoinError. Pondering that more.

Reply:
ah forget about the `JoinErr` :hankey: 

we don't actually need to spawn for each future, do we?  Just collect them all, then we can spawn once for `join_all`?  That'd give us a `Result<Iterator<bigtable::Result<usize>>>` that we can `result?.collect::<Result<Vec<usize>>()` on

Reply:
n/m, we're already async.  don't think we need to spawn at all?

Reply:
most of the speed up actually came from the spawn!

Reply:
idk exactly how join all works but guessing they were running on a single thread?

Reply:
@t-nelson , what does the sentence with the emoji mean, when it's at home?

Reply:
:fax: -> fax -> facts

Reply:
So: "I confirmed that we don't need to spawn since we're already async"?

Reply:
> idk exactly how join all works but guessing they were running on a single thread?

I don't think so.  I used this to test my theory, https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a012d7c439412a36c05acc696ccfd601

Reply:
> So: "I confirmed that we don't need to spawn since we're already async"?

i see that you are fluent in late-night-trent! yep!

Reply:
copying from discord DMs:

i think if we spawned in your code we wouldn't see the 0,1,2,etc. print out in order, doing that now
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=85e14e6f14930884483eeda7aa64980b

suggests to me the old code is running async but single threaded, where this code running async and multithreaded.

another reason i hate async 👿

Reply:
right, but being as it's IO bound ops, the value of multi-threaded vs. single-thread + multiplexed should be nil so long as everyone is out of each other's way. i suspect that any gains from multi-threading are hiding some other issue

Reply:
everytime we upload we try 3 different compression methods and choose the best. there's our cpu bound